apax/data/input_pipeline.py

```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
+            "box": np.zeros((n_samples, 3, 3), dtype=np.float32),
         }
 
         labels = {
-            "energy": [],
+            "energy": np.zeros(n_samples, dtype=np.float64),
         }
 
         if self.forces:
-            labels["forces"] = []
-        if self.stress:
-            labels["stress"] = []
-
-        for sample in samples:
-            inp, lab = sample
-
-            inputs["numbers"].append(inp["numbers"])
-            inputs["n_atoms"].append(inp["n_atoms"])
-            inputs["positions"].append(inp["positions"])
-            inputs["box"].append(inp["box"])
-            idx, offsets = compute_nl(inp["positions"], inp["box"], self.cutoff)
-            inputs["idx"].append(idx)
-            inputs["offsets"].append(offsets)
-
-            labels["energy"].append(lab["energy"])
-            if self.forces:
-                labels["forces"].append(lab["forces"])
-            if self.stress:
-                labels["stress"].append(lab["stress"])
+            labels["forces"] = np.zeros((n_samples, max_atoms, 3), dtype=np.float64)
 
-        max_atoms = np.max(inputs["n_atoms"])
-        max_nbrs = np.max([idx.shape[1] for idx in inputs["idx"]])
+        idxs = []
+        offsets = []
+        for i, (inp, lab) in enumerate(samples):
+            inputs["numbers"][i, : inp["n_atoms"]] = inp["numbers"]
+            inputs["n_atoms"][i] = inp["n_atoms"]
+            inputs["positions"][i, : inp["n_atoms"]] = inp["positions"]
+            inputs["box"][i] = inp["box"]
 
-        max_atoms = next_power_of_two(max_atoms)
-        max_nbrs = next_power_of_two(max_nbrs)
+            idx, offset = compute_nl(inp["positions"], inp["box"], self.cutoff)
+            idxs.append(idx)
+            offsets.append(offset)
 
-        for i in range(len(inputs["n_atoms"])):
-            inputs["idx"][i], inputs["offsets"][i] = pad_nl(
-                inputs["idx"][i], inputs["offsets"][i], max_nbrs
-            )
+            labels["energy"][i] = lab["energy"]
+            if self.forces:
+                labels["forces"][i, : inp["n_atoms"]] = lab["forces"]
 
-            zeros_to_add = max_atoms - inputs["numbers"][i].shape[0]
-            inputs["positions"][i] = np.pad(
-                inputs["positions"][i], ((0, zeros_to_add), (0, 0)), "constant"
-            )
-            inputs["numbers"][i] = np.pad(
-                inputs["numbers"][i], (0, zeros_to_add), "constant"
-            ).astype(np.int16)
+        max_nbrs = np.max([idx.shape[1] for idx in idxs])
+        max_nbrs = next_power_of_two(max_nbrs)  # TODO better intervals
+        max_nbrs = round_up_to_multiple(max_nbrs, self.nl_padding)
 
-            if "forces" in labels:
-                labels["forces"][i] = np.pad(
-                    labels["forces"][i], ((0, zeros_to_add), (0, 0)), "constant"
-                )
+        inputs["idx"] = np.zeros((n_samples, 2, max_nbrs), dtype=np.int16)
```

Comment:
int16 goes to `32767` does that mean structures with more than that can not be trained on? That is not relevant for DFT but if we ever consider coarse grained simulations we might want to keep that in mind? But I am also not entirely shure what `idx` from the nl actually is.

Reply:
That is true. I switched it out to int16 at some point because it cut down memory requirements by like 15 % or so. This was during the last update to the input pipeline when we tried to figure out how to train on large datasets.
I am not sure it will be relevant for training on coarse grained structures, as 33 K atoms might be too large of a structure for a GPU to handle.

idx is the sparse [2,n_neighbors] neighborlist.